src/plugins/layout.ts

```diff
@@ -73,201 +68,153 @@ export default class Layout implements IPlugin {
-  private onDomDiscoverComplete() {
-    this.domDiscoverComplete = true;
-    for (let i = 0; i < this.domPreDiscoverMutations.length; i++) {
-      this.processMultipleNodeEvents(this.domPreDiscoverMutations[i]);
-    }
+  // Add node to the ShadowDom to store initial adjacent node info in a layout and obtain an index
+  private discoverNode(node: Node, shadowDom: ShadowDom): ILayoutState {
+    shadowDom.insertShadowNode(node, getNodeIndex(node.parentNode), getNodeIndex(node.nextSibling));
+    return createLayoutState(node, shadowDom);
   }
 
   private processMultipleNodeEvents<T extends ILayoutEventInfo>(eventInfos: T[]) {
-    let eventsData: IEventData[] = [];
+    let eventsData: IEventInfo[] = [];
     for (let i = 0; i < eventInfos.length; i++) {
-      let eventState = this.createEventState(eventInfos[i]);
-      eventsData.push({
-        type: this.eventName,
-        state: eventState
-      });
-      this.layoutStates[eventState.index] = eventState;
+      let eventData = this.createEvent(eventInfos[i]);
+      if (eventData) {
+        eventsData.push({
+          origin: Origin.Layout,
+          type: eventInfos[i].action,
+          data: eventData,
+        });
+      }
     }
     addMultipleEvents(eventsData);
   }
 
-  private createEventState<T extends ILayoutEventInfo>(eventInfo: T): ILayoutState {
-    let node = eventInfo.node;
-    let layoutState: ILayoutState = createLayoutState(node, this.shadowDom);
-
+  private createEvent<T extends ILayoutEventInfo>(eventInfo: T): ILayoutEventData {
+    let event: ILayoutEventData = null;
+    let index = eventInfo.index;
     switch (eventInfo.action) {
       case Action.Insert:
+        let insertedElement = eventInfo.node as Element;
+        let insertEvent = event = EventProvider.createInsert(insertedElement, this.shadowDom, this.mutationSequence);
+        this.states.add(insertEvent.state);
         // Watch element for scroll and input change events
-        this.watch(node, layoutState);
-        layoutState.action = Action.Insert;
-        break;
-      case Action.Update:
-        // Watch element for scroll and input change events
-        this.watch(node, layoutState);
-        layoutState.action = Action.Update;
+        this.watch(insertedElement);
         break;
       case Action.Remove:
         // Index is passed explicitly because indices on removed nodes are cleared,
         // so at this point we can't obtain node's index from the node itself
-        layoutState.index = eventInfo.index;
-        layoutState.action = Action.Remove;
+        event = EventProvider.createRemove(index, this.mutationSequence);
         break;
       case Action.Move:
-        layoutState.action = Action.Move;
+        let movedNode = eventInfo.node;
+        event = EventProvider.createMove(movedNode, this.mutationSequence);
+        break;
+      case Action.AttributeUpdate:
```

Comment:
>AttributeUpdate [](start = 18, length = 15)

Why do we need to know about this? When we didn't care about it before? Earlier we were looking for Updates - regardless of CharacterData update or Attributes? We should not have sub-event for different events, we should one format for layout updates.. everything else should be handled by compression layer (for saving on bytes).